File: 21/direct/main.go
package main

import (
	"fmt"
	"time"
)

func main() {
	r1, r5 := 0, 0
	prev := 0
	seen := make(map[int]struct{})

	start := time.Now()

	for {
		r1 = 123
		if (r1 & 456) == 72 {
			break
		}
	}

	r1 = 0

done:
	for {
		prev = r1
		r5 = r1 | 65536
		r1 = 8586263

		for {
			// fmt.Printf("r1: %d, r5: %d\n", r1, r5)
			r1 = (((r1 + (r5 & 255)) & 16777215) * 65899) & 16777215
			if 256 > r5 {
				if len(seen) == 0 {
					// first termination: dump the value of r1
					fmt.Printf("Part 1 in %v: %d\n", time.Since(start), r1)
				}
				if _, ok := seen[r1]; ok {
					// First duplicate: dump the previous value of r1
					fmt.Printf("Part 2 in %v: %d\n", time.Since(start), prev)
					break done
				} else {
					seen[r1] = struct{}{}
				}
				break
			} else {
				// r2 = 0
				// for {
				// 	r3 = r2 + 1
				// 	r3 *= 256
				// 	if r3 > r5 {
				// 		r5 = r2
				// 		break
				// 	} else {
				// 		r2++
				// 	}
				// }
				r5 = r5 / 256
			}
		}
	}
}

File: 21/direct/src/main.rs
use std::collections::HashSet;
use std::time::Instant;

pub fn main() {
    let start = Instant::now();
    let n: u64 = 16777215;

    let mut prev_r1;
    let mut r1;
    let mut r5;
    let mut seen = HashSet::new();

    loop {
        r1 = 123;
        if (r1 & 456) == 72 {
            break
        }
    }

    r1 = 0;

'done:
    loop {
        prev_r1 = r1;

        r5 = prev_r1 | 65536;
        r1 = 8586263;

        loop {
            r1 = (((r1 + (r5 & 255))& n) * 65899) & n;

            if 256 > r5 {
                if seen.is_empty() {
                    println!("Part 1 in {}s: {}", elapsed_str(start.elapsed()), r1);
                }
                if !seen.insert(r1) {
                    println!("Part 2 in {}s: {}", elapsed_str(start.elapsed()), prev_r1);
                    break 'done;
                }
                break;
            } else {
                r5 = r5 / 256;
            }
        }
    }
}

fn elapsed_str(elapsed: std::time::Duration) -> f64 {
    return (elapsed.as_secs() as f64) + (elapsed.subsec_nanos() as f64 / 1000_000_000.0);
}
